lib/asciidoctor/reader.rb

```diff
@@ -837,10 +837,42 @@ def preprocess_include_directive raw_target, raw_attributes
         warn %(asciidoctor: ERROR: #{line_info}: maximum include depth of #{@maxdepth[:rel]} exceeded)
         return
       elsif ::RUBY_ENGINE_OPAL && ::JAVASCRIPT_IO_MODULE == 'xmlhttprequest'
-        # NOTE resolves uri relative to currently loaded document
-        # NOTE we defer checking if file exists and catch the 404 error if it does not
-        target_type = :file
-        inc_path = relpath = @include_stack.empty? && ::Dir.pwd == @document.base_dir ? target : %(#{@dir}/#{target})
+        # NOTE the only way to check if the file exists is to catch a 404 response in the IO module
+        target, target_type, base_dir = ((@path_resolver ||= PathResolver.new '\\').posixify target), :file, @document.base_dir
+        if target.start_with? 'file://'
+          inc_path = relpath = target
+        elsif Helpers.uriish? target
+          # NOTE only require allow-uri-read if cross origin
+          unless (@path_resolver.descends_from? target, base_dir) || (@document.attributes.key? 'allow-uri-read')
+            replace_next_line %(link:#{target}[])
+            return true
+          end
+          inc_path = relpath = target
+        elsif @path_resolver.absolute_path? target
+          inc_path = relpath = ((target.start_with? '/') ? 'file://' : 'file:///') + target
+        elsif (ctx_dir = (top_level = @include_stack.empty?) ? base_dir : @dir) == '.'
```

Comment:
In my opinion this test (+ assignment) is hard to read.
We are covering the case when the target is relative and we are on the top level include (ie. not on a nested include), that's right ?

Reply:
This assignment allows us to handle both a top-level include and a nested include in all remaining conditions.

Reply:
...and then it immediately tests the `.` condition.